src/lib/optimizer/strategy/column_pruning_rule.cpp

```diff
@@ -216,12 +216,16 @@ void recursively_gather_required_expressions(
     // Make sure the entry in required_expressions_by_node exists, then insert all expressions that the current node
     // needs
     auto& required_expressions_for_input = required_expressions_by_node[input];
+    const auto& expressions_provided_by_input = input->output_expressions();
     for (const auto& required_expression : required_expressions) {
       // Add the columns needed here (and above) if they come from the input node. Reasons why this might NOT be the
       // case are: (1) The expression is calculated in this node (and is thus not available in the input node), or
       // (2) we have two input nodes (i.e., a join) and the expressions comes from the other side.
-      if (input->find_column_id(*required_expression)) {
-        required_expressions_for_input.emplace(required_expression);
+      for (auto column_id = ColumnID{0}; column_id < expressions_provided_by_input.size(); ++column_id) {
+        if (*expressions_provided_by_input[column_id] == *required_expression) {
+          required_expressions_for_input.emplace(required_expression);
+          break;
+        }
```

Comment:
Isn't that kind of reimplementing `find_column_id()`?

Reply:
It is 1:1 `find_column_id()`. But calling `find_column_id()` results in fetching the `output_expressions()` of the desired input for each required expression. Imagine having a `SELECT` statement with 10 columns, then we would fetch the output expressions 10 times for any input (plus join columns, group by columns, etc.).

Reply:
@dey4ss, any comment on that?

Reply:
In general, fetching the output expressions once and loop over them rather than calling `find_column_id()` multiple times could be done anywhere. As most of the overhead of `find_column_id()` is reduced by this PR, I am not sure whether this would have measurable performance implications. Yet, I think adding such a function that takes a lambda as argument, comparably to `visit_[lqp/pqp/expression]`, can avoid future problems with using the output expression (as I am not the first one having performance issues at that point).

Reply:
Would you prefer either
```c++
iterate_output_expressions(const std::shared_ptr<Node>& lqp_node, Visitor visitor)
```
that iterates over the output expressions and calls the visitor for each output expression with the output expression as argument or 
```c++
match_output_expression(const std::shared_ptr<Node>& lqp_node, const AbstractExpression& expression, Visitor visitor)
```
that calls the visitor (without an argument) only when the output expression equals the provided expression?

Reply:
While applying the utility function to the mentioned code parts, I realized that we are super wasteful calling `output_expressions()` in general.
Example from `CardinalityEstimator::estimate_alias_node()`:
```c++
  auto column_statistics =
      std::vector<std::shared_ptr<BaseAttributeStatistics>>{alias_node.output_expressions().size()};

  for (size_t expression_idx{0}; expression_idx < alias_node.output_expressions().size(); ++expression_idx) {
    const auto& expression = *alias_node.output_expressions()[expression_idx];
    const auto input_column_id = alias_node.left_input()->get_column_id(expression);
    column_statistics[expression_idx] = input_table_statistics->column_statistics[input_column_id];
  }
```
We use `output_expresssions()` as if it was accessing a member variable rather than a function call in some cases. Across the code base, I found multiple occurrences of calling them _inside_ a loop to access each element with the `[]` operator, repeated accesses to obtain the size, etc. For now, I assume that besides fixing repeated calls of `[find|get]_column_id`, there's still something more to do.

Reply:
Good catch! The call in the loop header of `CardinalityEstimator::estimate_alias_node()` is even worse: we might call `output_expressions()` and `size()` for every loop iteration. Such style should be avoided according to our contribution guidelines:

> If you have to use old-style loops, keep in mind that the loop condition is evaluated every time: Instead of for (auto offset = size_t{0}; offset < something.size(); ++offset), the size should be retrieved just once. See also [this document](http://llvm.org/docs/CodingStandards.html#don-t-evaluate-end-every-time-through-a-loop).

Actually, the part regarding `size()` also applies to `iterate_output_expressions`.

Reply:
> Actually, the part regarding `size()` also applies to `iterate_output_expressions`.

Totally agree, gonna cache the size.  Out of curiosity, as `output_expressions` is a vector, shouldn't accessing its `.size()`  be constant?

Reply:
Yes, it should be. Still, it could be an additional method call and we have seen differences in performance when caching the size.